### PR TITLE
[SPARK-48211][SQL] DB2: Read SMALLINT as ShortType

### PR DESCRIPTION
**Pull request by yaooqinn**
### What changes were proposed in this pull request?

This PR supports read SMALLINT from DB2 as ShortType

### Why are the changes needed?

- 15 bits is sufficient
- we write ShortType as SMALLINT
- we read smallint from other builtin jdbc sources as ShortType

### Does this PR introduce _any_ user-facing change?

yes, we add a migration guide for this
### How was this patch tested?

changed tests

### Was this patch authored or co-authored using generative AI tooling?

no

### CODE REVIEW

File: docs/sql-migration-guide.md [hunk 1]
```diff
@@ -50,6 +50,7 @@ license: |
 - Since Spark 4.0, Oracle JDBC datasource will write TimestampType as TIMESTAMP WITH LOCAL TIME ZONE, while in Spark 3.5 and previous, write as TIMESTAMP. To restore the previous behavior, set `spark.sql.legacy.oracle.timestampMapping.enabled` to `true`.
 - Since Spark 4.0, MsSQL Server JDBC datasource will read TINYINT as ShortType, while in Spark 3.5 and previous, read as IntegerType. To restore the previous behavior, set `spark.sql.legacy.mssqlserver.numericMapping.enabled` to `true`.
 - Since Spark 4.0, MsSQL Server JDBC datasource will read DATETIMEOFFSET as TimestampType, while in Spark 3.5 and previous, read as StringType. To restore the previous behavior, set `spark.sql.legacy.mssqlserver.datetimeoffsetMapping.enabled` to `true`.
+- Since Spark 4.0, DB2 JDBC datasource will read SMALLINT as ShortType, while in Spark 3.5 and previous, it was read as IntegerType. To restore the previous behavior, you can cast the column to the old type.
```
Comment (dongjoon-hyun) [inline docs/sql-migration-guide.md:53]: Hmm, it means we need to edit SQL statement, right?

Comment (yaooqinn) [reply docs/sql-migration-guide.md:53]: Yes, do we need a legacy conf for this like `spark.sql.legacy.mssqlserver.numericMapping.enabled`

File: connector/docker-integration-tests/src/test/scala/org/apache/spark/sql/jdbc/DB2IntegrationSuite.scala [hunk 2]
```diff
@@ -92,7 +92,7 @@ class DB2IntegrationSuite extends DockerJDBCIntegrationSuite {
     assert(types(7).equals("class java.math.BigDecimal"))
     assert(types(8).equals("class java.math.BigDecimal"))
     assert(types(9).equals("class java.math.BigDecimal"))
-    assert(rows(0).getInt(0) == 17)
+    assert(rows(0).getShort(0) == 17)
```
Comment (dongjoon-hyun) [inline connector/docker-integration-tests/src/test/scala/org/apache/spark/sql/jdbc/DB2IntegrationSuite.scala:95]: Thank you for updates. Shall we keep a test coverage for legacy?

**Review by dongjoon-hyun — APPROVED** [review]
LGTM (except one comment about legacy behavior test coverage)

**Comment by yaooqinn** [comment]
Thank you @dongjoon-hyun for the help

**Review by dongjoon-hyun — APPROVED** [review]

**Comment by yaooqinn** [comment]
Merged to master

**Pull request closed without merging**